---
title: "Cellranger Tutorial on p12"
author: "[LabEx CORTEX Bioinformatics Platform](https://labex-cortex.universite-lyon.fr/) - Lyon 1 University - vignette produced by Emeric Texeraud"
date: "`r Sys.Date()`"
output: html_notebook
---


Cell Ranger is a set of analysis pipelines that process Chromium single-cell data to align reads, generate feature-barcode matrices, perform clustering and other secondary analysis, and more. Cell Ranger includes four pipelines relevant to the 3' Single Cell Gene Expression Solution and related products:

- cellranger **mkfastq** demultiplexes raw base call (BCL) files generated by Illumina sequencers into FASTQ files. It is a wrapper around Illumina's bcl2fastq, with additional features that are specific to 10x libraries and a simplified sample sheet format.

- cellranger **count** takes FASTQ files from cellranger mkfastq and performs alignment, filtering, barcode counting, and UMI counting. It uses the Chromium cellular barcodes to generate feature-barcode matrices, determine clusters, and perform gene expression analysis. The count pipeline can take input from multiple sequencing runs on the same GEM well. cellranger count also processes Feature Barcode data alongside Gene Expression reads.

- cellranger **aggr** aggregates outputs from multiple runs of cellranger count, normalizing those runs to the same sequencing depth and then recomputing the feature-barcode matrices and analysis on the combined data. The aggr pipeline can be used to combine data from multiple samples into an experiment-wide feature-barcode matrix and analysis.

- cellranger **reanalyze** takes feature-barcode matrices produced by cellranger count or cellranger aggr and reruns the dimensionality reduction, clustering, and gene expression algorithms using tunable parameter settings.

- cellranger **multi** is used to analyze Cell Multiplexing data. It inputs FASTQ files from cellranger mkfastq and performs alignment, filtering, barcode counting, and UMI counting. It uses the Chromium cellular barcodes to generate feature-barcode matrices, determine clusters, and perform gene expression analysis. The cellranger multi pipeline also supports the analysis of Feature Barcode data.


More information can be found [here](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger).

Cellranger is a linux command-line tool, thus not used in R but only in a linux terminal.
Firstly, you need to access the directory where the output files from Cellranger will be kept. 
```{bash}
cd '/mnt/Data1/public_vignettes/cellranger_data'
```

Then you can start cellranger counting, specifying where the input directory (directory which contains the fastq files) is.
```{bash}
/opt/cellranger/cellranger-6.1.1/cellranger count --id=Nx1 --fastqs=/mnt/Data2/fastq/pilot --transcriptome=/opt/cellranger/transcriptome/refdata-gex-mm10-2020-A --localcores=32 --localmem=48
```

arguments : 
--id specify the name of the directory created by Cellranger
--fastqs specify in which directory are the fastqs
--transcriptome specify where is the transcriptome (you can verify that you have the [latest transcriptome](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest))
--localcores specify the number of cores you want to use for parallelization
--localmem specify how much RAM memory you can spend for the count process 

In our case, we also have a second experiment we need to count : 
```{bash}
/opt/cellranger/cellranger-6.1.1/cellranger count --id=Nx2 --fastqs=/mnt/Data2/fastq/Nx --transcriptome=/opt/cellranger/transcriptome/refdata-gex-mm10-2020-A --localcores=32 --localmem=48
```

We can also do all these lines in one single line, adding `shutdown -h now` to tell the computer to shutdown after the jobs are done.

```{bash cellranger count, eval=FALSE}
cd '/mnt/Data1/Raineteau/2020_03_10X_dSVZ_p12_Nx_Hx/cellranger_count'; /opt/cellranger/cellranger-6.1.1/cellranger count --id=Nx1_v6 --fastqs=/mnt/Data2/Raineteau/2020_03_10X_dSVZ_p12_Nx_Hx/pilot --transcriptome=/opt/cellranger/transcriptome/refdata-gex-mm10-2020-A --localcores=32 --localmem=48; /opt/cellranger/cellranger-6.1.1/cellranger count --id=Nx2_v6 --fastqs=/mnt/Data2/Raineteau/2020_03_10X_dSVZ_p12_Nx_Hx/Nx --transcriptome=/opt/cellranger/transcriptome/refdata-gex-mm10-2020-A --localcores=32 --localmem=48; shutdown -h now
```
