---
title: "Cellranger tutorial on public data"
author: "[LabEx CORTEX Bioinformatics Platform](https://labex-cortex.universite-lyon.fr/) - Lyon 1 University - vignette produced by Emeric Texeraud"
date: "`r Sys.Date()`"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Cell Ranger <a name="cellranger"></a>

Cell Ranger is a set of analysis pipelines that process Chromium single-cell data to align reads, generate feature-barcode matrices, perform clustering and other secondary analysis, and more. Cell Ranger includes four pipelines relevant to the 3' Single Cell Gene Expression Solution and related products:

- cellranger **mkfastq** demultiplexes raw base call (BCL) files generated by Illumina sequencers into FASTQ files. It is a wrapper around Illumina's bcl2fastq, with additional features that are specific to 10x libraries and a simplified sample sheet format.

- cellranger **count** takes FASTQ files from cellranger mkfastq and performs alignment, filtering, barcode counting, and UMI counting. It uses the Chromium cellular barcodes to generate feature-barcode matrices, determine clusters, and perform gene expression analysis. The count pipeline can take input from multiple sequencing runs on the same GEM well. cellranger count also processes Feature Barcode data alongside Gene Expression reads.

- cellranger **aggr** aggregates outputs from multiple runs of cellranger count, normalizing those runs to the same sequencing depth and then recomputing the feature-barcode matrices and analysis on the combined data. The aggr pipeline can be used to combine data from multiple samples into an experiment-wide feature-barcode matrix and analysis.

- cellranger **reanalyze** takes feature-barcode matrices produced by cellranger count or cellranger aggr and reruns the dimensionality reduction, clustering, and gene expression algorithms using tunable parameter settings.

- cellranger **multi** is used to analyze Cell Multiplexing data. It inputs FASTQ files from cellranger mkfastq and performs alignment, filtering, barcode counting, and UMI counting. It uses the Chromium cellular barcodes to generate feature-barcode matrices, determine clusters, and perform gene expression analysis. The cellranger multi pipeline also supports the analysis of Feature Barcode data.


More information can be found [here](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger).



### Installation

First, add Cell Ranger to your PATH, by writing the following command in your `.bashrc` file (located in your home directory). Here we installed Cell Ranger in the `/opt` directory :
```{bash cellranger_path, eval=FALSE}
export PATH=/opt/cellranger/cellranger-5.0.1/bin:$PATH
```
This will allow you to call cellranger without typing the entire path.
To verify the good installation, open a terminal and type `cellranger`. If `command not found`, then Cell Ranger was not added correctly to your PATH (The main reason could be that the path you provided is not correct). More info about installation can be found [here](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/installation).

Download the latest transcriptome [here](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest).

### Count <a name="count"></a>
Before running the cellranger `count` on the single cell RNA-seq data, there is a need to demultiplex the signals. It is not part of our pipeline, as the data are usually demultiplexed by the sequencing platform. To know more about the demultiplexing command in Cell Ranger, see the [mkfastq tutorial](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/mkfastq). The output files are called fastqs. 6 fastq files are generated per library.


The count process is time consuming, usually taking several hours for big datasets. It takes as minimal parameters the id you want to give to the data (name of the output directory/folder), the path to the directory containing the fastq files, and the path to the transcriptome of the appropriate species. Optionally, the number of cores and memory size used can be specified, as shown in example :
```{bash cellranger count, eval=FALSE}
cellranger count --id=neuron_10k --fastqs=/mnt/Data1/public_vignettes/dataset/neuron_10k_v3_fastqs --transcriptome=/opt/cellranger/transcriptome/refdata-gex-mm10-2020-A --localcores=32 --localmem=48
```

If you want to power off the computer automatically at the end of the task, add `; shutdown -h now` :
```{bash cellranger_count_shutdown, eval=FALSE}
cellranger count --id=neuron_10k --fastqs=/mnt/Data1/public_vignettes/dataset/neuron_10k_v3_fastqs --transcriptome=/opt/cellranger/transcriptome/refdata-gex-mm10-2020-A --localcores=32 --localmem=48; shutdown -h now
```


